Don't use globals for platforms and components registration

invidian opened the pull request:
~**This pull request includes commits from #1098, please do not review commits from #1098 in this pull request**~

Closes #992
Closes #597

See commit messages for details.

knrt10 commented on pkg/platform/aws/aws.go:
```diff
@@ -88,6 +88,11 @@ type config struct {
+
 // init registers aws as a platform
 func init() {
 	platform.Register("aws", NewConfig())
```
Can you please use the constant here also? Also, why not merge this commit and aks one?

```suggestion
platform.Register(Name, NewConfig())
```

knrt10 replied:
Now that I think of it, as you removed init function in the next commit. I would say it is un-necessary. But I would suggest merging this commit with AKS one.

invidian replied:
Yes, I should've done this. Will fix regardless.

knrt10 commented on pkg/platform/baremetal/baremetal.go:
```diff
@@ -60,6 +60,11 @@ type config struct {
 }
 
+const (
+	// Name represents bare-metal platform name as it should be referenced in function calls and configuration.
```
We should be consistent like other platforms.

```suggestion
      // Name represents Bare-metal platform name as it should be referenced in function calls and configuration.
```

invidian replied:
True, but it should be `Bare Metal`, as in quickstart guide.

knrt10 commented on pkg/platform/baremetal/baremetal.go:
```diff
@@ -60,6 +60,11 @@ type config struct {
+
 // init registers bare-metal as a platform
 func init() {
 	platform.Register("bare-metal", NewConfig())
```
Same

knrt10 commented on pkg/platform/packet/packet.go:
```diff
@@ -102,6 +102,11 @@ type config struct {
+
 // init registers packet as a platform
 func init() {
 	platform.Register("packet", NewConfig())
```
Same

knrt10 commented on pkg/components/cluster-autoscaler/component.go:
```diff
@@ -118,7 +118,10 @@ type packetConfiguration struct {
 }
 
-func newComponent() *component {
+// NewConfig returns new Cluster Autoscaler component configuration with default values set.
```
We should be consistent with the component name as used in the previous commit.

```suggestion
// NewConfig returns new cluster-autoscaler component configuration with default values set.
```

invidian replied:
Ditto official name: https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#cluster-autoscaler

invidian replied:
Fixed.

knrt10 commented on pkg/components/external-dns/component.go:
```diff
@@ -88,7 +88,10 @@ type component struct {
 }
 
-func newComponent() *component {
+// NewConfig returns new ExternalDNS component configuration with default values set.
```
```suggestion
// NewConfig returns new external-dns component configuration with default values set.
```

invidian replied:
`ExternalDNS` is the official form of the project name: https://github.com/kubernetes-sigs/external-dns#externaldns.

knrt10 replied:
Cool, then it needs to be fixed for the previous commit then

invidian replied:
Right, fixing.

knrt10 commented on pkg/components/linkerd/component.go:
```diff
@@ -56,7 +56,10 @@ type cert struct {
 }
 
-func newComponent() *component {
+// NewConfig returns new Istio component configuration with default values set.
```
This is Linkerd 🙈 

```suggestion
// NewConfig returns new Linkerd component configuration with default values set.
```

invidian replied:
Nice catch, fixed, thanks!

knrt10 reviewed (changes requested):
Awesome work 🥳 . Some suggestions.

surajssd commented on cli/cmd/cluster/cluster.go:
```diff
@@ -30,12 +30,6 @@ import (
 	"github.com/kinvolk/lokomotive/pkg/platform"
```
> All platform packages are now imported by utils.go file, so anonymous imports are no longer needed.

I am not sure if this commit message is valid. The anonymous imports are not needed anymore because we don't rely on the init function call registration method. So importing in `utils.go` or not does not matter.

invidian replied:
Good point. Fixed.

surajssd commented on test/components/install_test.go:
```diff
@@ -20,35 +20,13 @@ package components
-
-	"github.com/kinvolk/lokomotive/pkg/components"
-	_ "github.com/kinvolk/lokomotive/pkg/components/flatcar-linux-update-operator"
+	flatcarlinuxupdateoperator "github.com/kinvolk/lokomotive/pkg/components/flatcar-linux-update-operator"
```
fluo in short?

surajssd commented on pkg/components/dex/component_test.go:
```diff
@@ -17,7 +17,7 @@ package dex_test
 import (
```
Typo in commit message: `realy`.

surajssd reviewed (changes requested):
We also need to parallelize the tests.

invidian commented on pkg/platform/aws/aws.go:
```diff
@@ -88,9 +88,14 @@ type config struct {
 	KubeAPIServerExtraFlags  []string
 }
 
+const (
```
Commit as doing different things. One is exposing things, other is adding new exported const. Both commit still do one thing each, just in multiple contexts. I think I prefer to keep them separate.

invidian commented:
>We also need to parallelize the tests.

Which ones? I'm almost sure this is out of scope of this (already large) PR.

If you talk about the unit tests using configuration, I think we should fix them either on the way OR in a separate PR as part of other issue.

@surajssd addressed remaining review comments, please have a look again.

surajssd commented:
> If you talk about the unit tests using configuration, I think we should fix them either on the way OR in a separate PR as part of other issue.

The promise of this change is that tests could be parallelised which weren't because we were extracting the same object from global store. So how do we confirm that promise is fullfilled with this change?

invidian commented:
>The promise of this change is that tests could be parallelised which weren't because we were extracting the same object from global store. So how do we confirm that promise is fullfilled with this change?

There is no global registry anymore and it's up to the package itself to ensure immutability, as `NewConfig()` function returns a pointer to configuration. Each package must ensure this pointer is unique, not taken from global, which is the case. It will be fine :)

invidian commented:
I need to rebase.

invidian commented:
Rebased to also adapt Tinkerbell platform for introduced changes.

iaguis commented on cli/cmd/cluster/component.go:
```diff
@@ -15,33 +15,74 @@
+	"fmt"
 
 	"github.com/kinvolk/lokomotive/pkg/components"
+	awsebscsidriver "github.com/kinvolk/lokomotive/pkg/components/aws-ebs-csi-driver"
```
ew, this looks ugly but I guess that's the way it is :D

invidian replied:
Looks like cat running over the keyboard, indeed :smile:

iaguis reviewed (approved):
LGTM. Thanks for getting rid of all those terrible `init()` functions!
